frontend/public/components/monitoring/receiver-forms/routing-labels-editor.tsx

```diff
@@ -0,0 +1,162 @@
+import * as _ from 'lodash-es';
+import * as React from 'react';
+import { MinusCircleIcon, PlusCircleIcon } from '@patternfly/react-icons';
+import { Button } from '@patternfly/react-core';
+import { ExternalLink, SectionHeading } from '../../utils';
+
+const DEFAULT_RECEIVER_LABEL = 'All (default receiver)';
+
+export const RoutingLabelEditor = ({ formValues, setFormValues, isDefaultReceiver }) => {
+  const setRouteLabel = (path: string, v: any): void => {
+    const labels = _.clone(formValues.routeLabels);
+    _.set(labels, path.split(', '), v);
+    setFormValues({
+      ...formValues,
+      routeLabels: labels,
+    });
+  };
+
+  const onRoutingLabelChange = (path: string): ((e) => void) => {
+    return (e) => setRouteLabel(path, e.target.value);
+  };
+
+  const onRoutingLabelRegexChange = (e, i: number): void => {
+    setRouteLabel(`${i}, isRegex`, e.target.checked);
+  };
+
+  const addRoutingLabel = (): void => {
+    setRouteLabel(`${formValues.routeLabels.length}`, {
+      name: '',
+      value: '',
+      isRegex: false,
+    });
+  };
+
+  const removeRoutingLabel = (i: number): void => {
+    const labels = _.clone(formValues.routeLabels);
+    labels.splice(i, 1);
+    setFormValues({
+      ...formValues,
+      routeLabels: labels,
```

Comment:
This might be an issue... I believe `useState` will batch updates like `setState` does, which means `formValues` might not be the latest value. This is one reason for `useReducer` instead of `useState`

Reply:
Ok, haven't seen an issue yet in testing or integration tests.  I'll look into `useReducer`

Reply:
ok, changes in this [commit](https://github.com/openshift/console/pull/3937/commits/5a2398fe5b22feb28a7974e7da50bad2dea8ade0)

Reply:
fixed